runtime/src/main/java/org/corfudb/runtime/view/SnowflakeGuidFactory.java

```diff
@@ -0,0 +1,49 @@
+package org.corfudb.runtime.view;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Fast Globally Unique Identity (GUID) generator that returns compact ids
+ * having a notion of a universal comparable ordering.
+ *
+ *    ______________________________________________
+ *   | timestamp lower 42 bits | Node Id | Sequence |
+ *    ------------------------- --------- ----------
+ *    <------ 42 bits ---------><-10bits-><-12bits-->
+ *
+ * Created by Sundar Sridharan on 5/22/19.
+ */
+public class SnowflakeGuidFactory implements AbstractOrderedGUIDFactory {
+    public static final short NODE_MAX_VAL = 1023;
+    public static final int NODE_BIT_POS = 10;
+    public static final short SEQ_MAX_VAL = 4095;
+    public static final int SEQ_MSB = 12;
+
+    private short node;
+    private AtomicInteger sequence;
+
+    /**
+     * Use this node id to scope the generator to a namespace.
+     * For example:
+     *      UUID nodeId = UUID.fromString("mac_address_of_machine");
+     *      SnowflakeGuidFactory guidFactory = new SnowflakeGuidFactory(nodeId.node());
+     * @param node
+     */
+    public SnowflakeGuidFactory(long node) {
+        this.node = (short)(node & NODE_MAX_VAL);
+        sequence = new AtomicInteger(0);
+    }
+
+    /**
+     * Fast best effort generation scheme for snowflakes generation.
+     * @return The next 64-bit integer.
+     */
+    @Override
+    public long nextGUID() {
```

Comment:
Is this safe to be accessed by multiple threads?

Reply:
It should be. The only local variable accessed is the sequence and that is updated Atomically.